dvd_info.1
'\" t
.\"     Title: dvd_info
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.20
.\"      Date: 2025-05-26
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "DVD_INFO" "1" "2025-05-26" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
dvd_info \- displays information about a DVD
.SH "SYNOPSIS"
.sp
\fBdvd_info\fP [\fIDVD PATH\fP] [\fIOPTIONS\fP]
.SH "DESCRIPTION"
.sp
Displays information about a DVD optical disc, image, or directory in
human\-readable, JSON, or formatted chapter outputs.
.sp
The default device is based on your operating system, and is the primary
optical drive.
.sp
DVDs can store two kinds of subtitles \(em VOBSUB, which are images, and closed
captioning, which are text. Only the VOBSUB subtitles are reported here.
.sp
DVDs may be authored by the provider to intentionally have bugs on the
structure that will cause problems during playback or ripping if the correct
track is not selected. \fBdvd_info\fP does its best to work around and flag these
tracks as invalid. See the JSON or verbose output formats to see which ones
are tagged.
.sp
DVDs may also have audio and subtitle streams marked as inactive as well. This
is normal, and usually won\(cqt show up at all in other DVD applications.
.sp
The ID of a DVD is a md5 hash of the IFOs on the disc.
.sp
The disc title is the volume name of the DVD\(cqs UDF. It will only appear if the
source is an ISO file or a physical disc.
.sp
DVDs support three audio codecs: ac3 (Dolby Digital), dts, and pcm. Dolby
Digital and DTS support up to 6 channels (5.1 surround sound). PCM supports
stereo only.
.sp
DVDs support multiple audio and subtitle languages. These are displayed in
output as a two letter string. Examples: de \- German, en \- English, fr \-
French, it \- Italian, etc.
.SH "OPTIONS"
.sp
\fB\-t, \-\-track\fP=\fITRACK\fP
.RS 4
Limit output to the selected track. By default, list all tracks.
.RE
.sp
\fB\-j, \-\-json\fP
.RS 4
Format output in JSON. All detailed information is included.
.RE
.sp
\fB\-v, \-\-video\fP
.RS 4
Display detailed information about video streams.
.RE
.sp
\fB\-a, \-\-audio\fP
.RS 4
Display detailed information about audio streams.
.RE
.sp
\fB\-s, \-\-subtitles\fP
.RS 4
Display detailed information about VOBSUB subtitles.
.RE
.sp
\fB\-c, \-\-chapters\fP
.RS 4
Display detailed information about chapters.
.RE
.sp
\fB\-d, \-\-cells\fP
.RS 4
Display detailed information about cells.
.RE
.sp
\fB\-x, \-\-all\fP
.RS 4
Display all optional detailed information.
.RE
.sp
\fB\-l, \-\-longest\fP
.RS 4
Only display the longest track.
.RE
.sp
\fB\-A, \-\-has\-audio\fP
.RS 4
Limit output to tracks that have audio streams.
.RE
.sp
\fB\-N, \-\-has\-alang\fP=\fILANGUAGE\fP
.RS 4
Limit output to tracks with a an audio track of two\-character language
code. For example, \fIen\fP, \fIes\fP, \fIfr\fP.
.RE
.sp
\fB\-S, \-\-has\-subtitles\fP
.RS 4
Limit output to tracks that have VOBSUB subtitles.
.RE
.sp
\fB\-G, \-\-has\-slang\fP=\fILANGUAGE\fP
.RS 4
Limit output to tracks with a subtitles off two\-character language
code. For example, \fIen\fP, \fIes\fP, \fIfr\fP.
.RE
.sp
\fB\-E, \-\-seconds\fP=\fISECONDS\fP
.RS 4
Limit output to tracks with a minimum length of number of seconds.
.RE
.sp
\fB\-M, \-\-minutes\fP=\fIMINUTES\fP
.RS 4
Limit output to tracks with a minimum length of number of minutes.
Cannot be used in combination with \fI\-\-seconds\fP.
.RE
.sp
\fB\-T, \-\-vts\fP=\fINUMBER\fP
.RS 4
Limit output to tracks in Video Title Set number.
.RE
.sp
\fB\-L, \-\-valid\fP
.RS 4
Don\(cqt display tracks that may be flagged as broken or invalid.
.RE
.sp
\fB\-i, \-\-id\fP
.RS 4
Display DVD unique identifier only.
.RE
.sp
\fB\-u, \-\-volume\fP
.RS 4
Display DVD UDF volume name only.
.RE
.sp
\fB\-g, \-\-xchap\fP
.RS 4
Display track chapters in export format suitable for mkvmerge(1).
.RE
.sp
\fB\-h, \-\-help\fP
.RS 4
Display help output.
.RE
.sp
\fB\-\-version\fP
.RS 4
Display version information.
.RE
.SH "SEE ALSO"
.sp
dvd_backup(1), dvd_copy(1), dvd_drive_status(1), dvd_player(1), dvd_rip(1), dvd_trip(1)
.SH "BUGS"
.sp
Please file bugs at \c
.URL "https://github.com/beandog/dvd_info/issues" "" ""
.SH "AUTHOR"
.sp
dvd_info was written by Steve Dibb \c
.MTO "steve.dibb\(atgmail.com" "" ""
and was started as a
clone of lsdvd(1).
.SH "RESOURCES"
.sp
Source code available at GitHub: \c
.URL "https://github.com/beandog/dvd_info" "" ""
.SH "COPYING"
.sp
Copyright (C) 2023 Steve Dibb. Free use of this software is granted under the
terms of the GNU General Public License, version 2 (GPL).

ChangeLog
1.18

* dvd_rip: remove help option to deinterlace video

1.17

* Fix logging of libdvdread to be less verbose, and output to stderr (bug #11,
  thanks to koorogi for reporting)
* Require libdvdread minimum version 6.1.0 for new DVD logging functionality
* Fix segfault on creating md5sum of dvdread id on newer systems with gcc
  (bug #12, thanks to yzu1103309 for patch)
* dvd_info: Add options for verbose and debugging output to display any errors
  from libdvdread
* Support FreeBSD 14.1 - see https://github.com/beandog/dvd_info/wiki
* Rename original dvd_rip which uses libmpv to dvd_trip
* dvd_trip: Set video codec to x264 by default when using Matroska
* dvd_trip: Set default audio codec to AAC, container to MP4
* dvd_trip: Use native ffmpeg AAC audio encoder
* dvd_trip: Add --start and --stop options
* dvd_trip: Fix setting faststart on mp4s
* dvd_trip: Encode same number of channels as source
* Add new dvd_rip program that uses ffmpeg to rip
* dvd_backup: Fix -i / --ifos to not copy VOBs, and always overwrite existing
  backup files
* dvd_player: fix playing at starting chapter (and range)
* dvd_player: deinterlace video by default using bwdif
* dvd_player: drop option to detelecine
* dvd_player, dvd_rip: fix stopping at wrong end point for chapters
* dvd_player, dvd_rip: if chapter ranges given are incorrect, don't fix for
  the user, and instead quit
* dvd_drive_status: use nonblock when opening device
* Fix building on msys2
* Include header to get nprocs on Linux, thanks to har7an for bug report

1.16

* Don't require libdvdcss to build
* Remove dvd_eject, doesn't work very well and requires libdvdcss
* Fix building without libmpv in configure script (Gentoo bug 919832)
* Rename dvd_trip to dvd_rip
* dvd_info: display # of blocks per track in JSON
* dvd_info: add option to narrow by audio or subtitle by language
* dvd_info: add option to display detailed VTS info
* dvd_copy: fix copying last block of a cell (bug #5, thanks to remenor and
  tremon015)
* dvd_rip: Rip to H.264 / MP4 by default

1.15

* Make display output more human-friendly

1.14

* dvd_player: fix bug where custom configuration file not loading

1.13

* Add dvd_debug, a wrapper to libdvdread's ifo_print()
* Add autotools scripts
* dvd_info: Fix typo in JSON output (thanks to gurpreetatwal)
* dvd_info: Add option to only display the longest track
* dvd_trip: Add detelecine and deinterlace as options

1.12

* Port to msys (thanks to veso266)
* dvd_copy: when given a single cell or chapter, don't use a range
* dvd_backup: allow setting DVD name for directory
* dvd_backup: add option to backup IFOs only

1.11

* Fix build with libdvdread-6.1.1

1.10

* dvd_backup: fix reading all title sets

1.9

* Add dvd_trip - a tiny little DVD ripper using libmpv as backend

1.8

* Update detelecine filters for dvd_player
* Require mpv version >= 0.29.0 for dvd_player

1.7

* Add dvd_player - a tiny little DVD player using libmpv

1.6 - 2019-06-08

* dvd_info: additional checks for invalid tracks
* dvd_info: fix dvd xchap output always returning the same track
* dvd_info: add verbose output option
* dvd_info: don't display invalid tracks, invalid streams, or tracks with
  length under one second by default
* dvd_info: add option to display tracks in video title set

1.5 - 2019-06-01

* Fix displaying longest track
* Fix build without man page error

1.4 - 2019-05-30

* Cleanup code for portability to other systems
* New man pages
* Proper boundary checks for user arguments

1.3 - 2019-05-30

* Detect intentionally broken DVDs and flag tracks as invalid and display
  errors

1.2 - 2018-02-28

* dvd_copy can copy track to stdout

1.1 - 2018-02-21

* Fixed a bug in dvd_copy where "Disc title" output would be garbage

1.0 - 2018-02-20

* First release after working on this for years, woots!
* dvd_info, dvd_copy, dvd_drive_status initial builds


dvd_info.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <stdbool.h>
#include <getopt.h>
#include "config.h"
#include "dvd_info.h"
#include "dvd_open.h"
#include "dvd_device.h"
#include "dvd_vmg_ifo.h"
#include "dvd_vts.h"
#include "dvd_track.h"
#include "dvd_chapter.h"
#include "dvd_cell.h"
#include "dvd_video.h"
#include "dvd_audio.h"
#include "dvd_subtitles.h"
#include "dvd_time.h"
#include "dvd_json.h"
#include "dvd_xchap.h"
#include "dvd_vob.h"
#include "dvd_init.h"
#ifdef __linux__
#include <linux/cdrom.h>
#include <linux/limits.h>
#include "dvd_drive.h"
#else
#include <limits.h>
#endif

	/**
	 *      _          _     _        __
	 *   __| |_   ____| |   (_)_ __  / _| ___
	 *  / _` \ \ / / _` |   | | '_ \| |_ / _ \
	 * | (_| |\ V / (_| |   | | | | |  _| (_) |
	 *  \__,_| \_/ \__,_|___|_|_| |_|_|  \___/
	 *                 |_____|
	 *
	 * ** display information about a DVD **
	 *
	 * dvd_info is a clone of lsdvd that adds additional features, such as
	 * JSON output, OGM chapter support, and checking for track errors.
	 *
	 */

int main(int argc, char **argv) {

	// Program name
	bool p_dvd_json = false;
	bool p_dvd_xchap = false;
	bool p_dvd_id = false;
	bool p_dvd_title = false;

	// lsdvd similar display output
	bool d_audio = false;
	bool d_video = false;
	bool d_chapters = false;
	bool d_subtitles = false;
	bool d_cells = false;
	bool d_title_sets = false;

	// How much output
	bool verbose = false;
	log_verbose = false;
	log_debug = false;

	// limit results
	bool d_has_audio = false;
	bool d_has_subtitles = false;
	bool d_has_alang = false;
	bool d_has_slang = false;
	char d_alang[DVD_AUDIO_LANG_CODE + 1] = {'\0'};
	char d_slang[DVD_AUDIO_LANG_CODE + 1] = {'\0'};
	bool d_longest = false;
	bool opt_min_seconds = true;
	unsigned long int arg_number = 0;
	uint32_t arg_min_seconds = 0;
	bool opt_min_minutes = true;
	uint32_t arg_min_minutes = 0;
	bool opt_vts = false;
	uint16_t arg_vts = 0;
	bool d_is_valid = false;

	// dvd_info
	bool d_disc_title_header = true;
	uint16_t d_first_track = 1;
	uint16_t d_last_track = 1;
	uint16_t track_number = 1;
	uint16_t vts = 1;
	uint8_t audio_track_ix = 0;
	uint8_t subtitle_track_ix = 0;
	uint8_t chapter_ix = 0;
	uint8_t cell_ix = 0;
	uint8_t d_stream_num = 0;

	// Device hardware
	char device_filename[PATH_MAX];

	// libdvdread
	dvd_reader_t *dvdread_dvd = NULL;
	ifo_handle_t *vmg_ifo = NULL;
	ifo_handle_t *vts_ifo = NULL;

	// DVD
	struct dvd_info dvd_info;

	// Video Title Set
	struct dvd_vts dvd_vts[DVD_MAX_VTS_IFOS];

	// Display formats
	const char *display_formats[4] = { "Pan and Scan or Letterbox", "Pan and Scan", "Letterbox", "Unset" };

	// getopt_long
	bool valid_args = true;
	bool opt_track_number = false;
	uint16_t arg_track_number = 0;
	int ix = 0;
	int opt = 0;
	bool invalid_opt = false;
	const char p_short_opts[] = "aAcdeE:gG:hijlLM:N:sST:t:uVvxyz";
	struct option p_long_opts[] = {

		{ "track", required_argument, NULL, 't' },
		{ "vts", required_argument, NULL, 'T' },

		{ "audio", no_argument, NULL, 'a' },
		{ "video", no_argument, NULL, 'v' },
		{ "chapters", no_argument, NULL, 'c' },
		{ "subtitles", no_argument, NULL, 's' },
		{ "cells", no_argument, NULL, 'd' },
		{ "title-sets", no_argument, NULL, 'e' },
		{ "all", no_argument, NULL, 'x' },

		{ "json", no_argument, NULL, 'j' },
		{ "xchap", no_argument, NULL, 'g' },
		{ "id", no_argument, NULL, 'i' },
		{ "volume", no_argument, NULL, 'u' },

		{ "longest", required_argument, NULL, 'l' },
		{ "min-seconds", required_argument, NULL, 'E' },
		{ "min-minutes", required_argument, NULL, 'M' },
		{ "has-audio", no_argument, NULL, 'A' },
		{ "has-alang", required_argument, NULL, 'N' },
		{ "has-subtitles", no_argument, NULL, 'S' },
		{ "has-slang", required_argument, NULL, 'G' },
		{ "valid", no_argument, NULL, 'L' },

		{ "help", no_argument, NULL, 'h' },
		{ "version", no_argument, NULL, 'V' },
		{ "verbose", no_argument, NULL, 'y' },
		{ "debug", no_argument, NULL, 'z' },
		{ 0, 0, 0, 0 }

	};

	// parse options
	while((opt = getopt_long(argc, argv, p_short_opts, p_long_opts, &ix)) != -1) {

		// It's worth noting that if there are unknown options passed,
		// I just ignore them, and continue printing requested data.
		switch(opt) {

			case 'a':
				d_audio = true;
				break;

			case 'A':
				d_has_audio = true;
				break;

			case 'c':
				d_chapters = true;
				break;

			case 'd':
				d_cells = true;
				break;

			case 'e':
				d_title_sets = true;
				break;

			case 'E':
				opt_min_seconds = true;
				arg_min_seconds = (uint32_t)strtoul(optarg, NULL, 10);
				break;

			case 'g':
				p_dvd_xchap = true;
				d_disc_title_header = false;
				d_chapters = true;
				break;

			case 'G':
				if(strlen(optarg) != 2 || !isalpha(optarg[0]) || !isalpha(optarg[1])) {
					fprintf(stderr, "Subtitle language code must be two characters\n");
					return 1;
				}
				strncpy(d_slang, optarg, DVD_AUDIO_LANG_CODE);
				d_has_slang = true;
				break;

			case 'i':
				p_dvd_id = true;
				break;

			case 'j':
				p_dvd_json = true;
				d_disc_title_header = false;
				d_audio = true;
				d_video = true;
				d_chapters = true;
				d_subtitles = true;
				d_cells = true;
				break;

			case 'l':
				d_longest = true;
				break;

			case 'L':
				d_is_valid = true;
				break;

			case 'M':
				opt_min_minutes = true;
				arg_min_minutes = (uint32_t)strtoul(optarg, NULL, 10);
				break;

			case 'N':
				if(strlen(optarg) != 2 || !isalpha(optarg[0]) || !isalpha(optarg[1])) {
					fprintf(stderr, "Audio language code must be two characters\n");
					return 1;
				}
				strncpy(d_alang, optarg, DVD_AUDIO_LANG_CODE);
				d_has_alang = true;
				break;

			case 's':
				d_subtitles = true;
				break;

			case 'S':
				d_has_subtitles = true;
				break;

			case 'T':
				opt_vts = true;
				arg_number = strtoul(optarg, NULL, 10);
				if(arg_number > 99)
					arg_vts = 99;
				else
					arg_vts = (uint16_t)arg_number;
				break;


			case 't':
				opt_track_number = true;
				arg_number = strtoul(optarg, NULL, 10);
				if(arg_number > 99)
					arg_track_number = 99;
				else if(arg_number == 0)
					arg_track_number = 1;
				else
					arg_track_number = (uint16_t)arg_number;
				break;

			case 'u':
				p_dvd_title = true;
				break;

			case 'v':
				d_video = true;
				break;

			case 'V':
				printf("dvd_info %s\n", PACKAGE_VERSION);
				return 0;

			case 'x':
				d_audio = true;
				d_video = true;
				d_chapters = true;
				d_subtitles = true;
				d_cells = true;
				break;

			case 'y':
				verbose = true;
				log_verbose = true;
				break;

			case 'z':
				verbose = true;
				log_verbose = true;
				log_debug = true;
				break;

			// ignore unknown arguments
			case '?':
				invalid_opt = true;
			case 'h':
				printf("dvd_info - display information about a DVD\n");
				printf("\n");
				printf("Usage: dvd_info [path] [options]\n");
				printf("\n");
				printf("Options:\n");
				printf("  -t, --track <number>  Limit to selected track (default: all tracks)\n");
				printf("  -j, --json            Display output in JSON format\n");
				printf("\n");
				printf("Detailed information:\n");
				printf("  -v, --video           Display video streams\n");
				printf("  -a, --audio           Display audio streams\n");
				printf("  -s, --subtitles       Display VobSub subtitles\n");
				printf("  -c, --chapters        Display chapters\n");
				printf("  -d, --cells           Display cells\n");
				printf("  -e, --title-sets      Display video title sets\n");
				printf("  -x, --all             Display all\n");
				printf("\n");
				printf("Narrow results:\n");
				printf("  -l, --longest		Track with the longest length\n");
				printf("  -A, --has-audio       Track has audio streams\n");
				printf("  -N, --has-alang <xx>  Track has audio audio language, two character code\n");
				printf("  -S, --has-subtitles   Track has VobSub subtitles\n");
				printf("  -G, --has-slang <xx>  Track has subtitle language, two character code\n");
				printf("  -E, --seconds <secs>  Track has minimum number of seconds\n");
				printf("  -M, --minutes <mins>  Track has minimum number of minutes\n");
				printf("  -T, --vts <number>    Track is in video title set number\n");
				printf("  -L, --valid		Track is marked as valid\n");
				printf("\n");
				printf("Other:\n");
				printf("  -i, --id		Display DVD ID only\n");
				printf("  -u, --volume		Display DVD UDF volume name only (for ISO or disc)\n");
				printf("  -g, --xchap           Display title's chapter format for mkvmerge\n");
				printf("  -h, --help            Display these help options\n");
				printf("  -v, --verbose         Display verbose output\n");
				printf("  -z, --debug           Display debugging output\n");
				printf("      --version         Display version\n");
				printf("\n");
				printf("DVD path can be a device name, a single file, or a directory (default: %s)\n", DEFAULT_DVD_DEVICE);
				if(invalid_opt)
					return 1;
				return 0;
			// let getopt_long set the variable
			case 0:
			default:
				break;

		}

	}

	// If '-i /dev/device' is not passed, then set it to the string
	// passed.  fex: 'dvd_info /dev/dvd1' would change it from the default
	// of '/dev/dvd'.
	memset(device_filename, '\0', PATH_MAX);
	if (argv[optind])
		strncpy(device_filename, argv[optind], PATH_MAX - 1);
	else
		strncpy(device_filename, DEFAULT_DVD_DEVICE, PATH_MAX - 1);

	// Exit after all invalid input warnings have been sent
	if(valid_args == false)
		return 1;

	/** Begin dvd_info :) */

	// Use a custom function to send logs to (and shut up the annoying ones)
	dvd_logger_cb dvdread_logger_cb = { dvd_info_logger_cb };

	// Open the DVD
	dvdread_dvd = DVDOpen2(NULL, &dvdread_logger_cb, device_filename);

	if(!dvdread_dvd) {
		fprintf(stderr, "Opening DVD %s failed\n", device_filename);
		return 1;
	}

	// Open VMG IFO -- where all the cool stuff is
	vmg_ifo = ifoOpen(dvdread_dvd, 0);
	if(vmg_ifo == NULL || vmg_ifo->vmgi_mat == NULL) {
		fprintf(stderr, "Opening VMG IFO failed, can not open DVD\n");
		DVDClose(dvdread_dvd);
		return 1;
	}

	dvd_info = dvd_info_open(dvdread_dvd, device_filename);
	if(dvd_info.valid == 0)
		return 1;

	// Display ID only if requested
	if(p_dvd_id) {
		printf("%s\n", dvd_info.dvdread_id);
		return 0;
	}

	// Display volume name only if requested
	if(p_dvd_title) {
		printf("%s\n", dvd_info.title);
		return 0;
	}

	// Exit if track number requested does not exist
	if(opt_track_number && (arg_track_number > dvd_info.tracks || arg_track_number < 1)) {
		fprintf(stderr, "Valid track numbers: 1 to %" PRIu16 "\n", dvd_info.tracks);
		ifoClose(vmg_ifo);
		DVDClose(dvdread_dvd);
		return 1;
	} else if(opt_track_number) {
		d_first_track = arg_track_number;
		d_last_track = arg_track_number;
		track_number = d_first_track;
	} else {
		d_first_track = 1;
		d_last_track = dvd_info.tracks;
	}

	if(opt_vts && (arg_vts == 0 || arg_vts > dvd_info.video_title_sets)) {
		fprintf(stderr, "Video Title Set must be between 1 and %" PRIu16 "\n", dvd_info.video_title_sets);
		return 1;
	}

	// Do some checks to see if a VTS is ok or not
	for(vts = 1; vts < dvd_info.video_title_sets + 1; vts++)
		dvd_vts[vts] = dvd_vts_open(dvdread_dvd, vts);

	/**
	 * Track information
	 */

	struct dvd_track *dvd_tracks;
	dvd_tracks = dvd_tracks_init(dvdread_dvd, vmg_ifo, d_audio, d_subtitles, d_chapters, d_cells);

	dvd_info.longest_track = dvd_tracks[0].track;

	// Only display the longest track if requested
	if(d_longest) {
		d_first_track = dvd_info.longest_track;
		d_last_track = dvd_info.longest_track;
	}

	/** JSON display output **/

	if(p_dvd_json) {
		dvd_json(dvd_info, dvd_tracks, track_number, d_first_track, d_last_track);
		goto cleanup;
	}

	// Start dvd_info output

	if(opt_track_number)
		d_disc_title_header = false;

	if(d_disc_title_header && !p_dvd_xchap) {
		printf("Disc title: '%s', ", dvd_info.title);
		printf("ID: '%s', ", dvd_info.dvdread_id);
		printf("Tracks: %" PRIu16 ", ", dvd_info.tracks);
		printf("Longest track: %" PRIu16, dvd_info.longest_track);
		printf("\n");
	}

	/** dvdxchap display output **/
	if(p_dvd_xchap) {
		if(opt_track_number)
			dvd_xchap(dvd_tracks[arg_track_number]);
		else
			dvd_xchap(dvd_tracks[dvd_info.longest_track]);
		goto cleanup;
	}

	// Count valid, invalid tracks and title sets
	if(verbose || opt_vts || d_title_sets) {

		for(ix = 1; ix <= dvd_info.tracks; ix++) {
			dvd_vts[dvd_tracks[ix].vts].tracks++;
			if(dvd_tracks[ix].valid) {
				dvd_info.valid_tracks++;
				dvd_vts[dvd_tracks[ix].vts].valid_tracks++;
			} else {
				dvd_info.invalid_tracks++;
				dvd_vts[dvd_tracks[ix].vts].invalid_tracks++;
			}
		}

	}

	// Print the valid and invalid VTSs
	if(verbose || d_title_sets) {

		printf("        Tracks: %*" PRIu16 ", ", 2, dvd_info.tracks);
		printf("Valid: %*" PRIu16 ", ", 2, dvd_info.valid_tracks);
		printf("Invalid: %*" PRIu16, 2, dvd_info.invalid_tracks);
		printf("\n");

		for(vts = 1; vts < dvd_info.video_title_sets + 1; vts++) {
			if(dvd_vts[vts].valid == true)
				dvd_info.valid_video_title_sets++;
			else
				dvd_info.invalid_video_title_sets++;
		}

		printf("        Video Title Sets: %*" PRIu16 ", ", 2, dvd_info.video_title_sets);
		printf("Valid: %*" PRIu16 ", ", 2, dvd_info.valid_video_title_sets);
		printf("Invalid: %*" PRIu16, 2, dvd_info.invalid_video_title_sets);
		printf("\n");

		for(vts = 1; vts < dvd_info.video_title_sets + 1; vts++) {
			printf("        Title set: %*" PRIu16 ", ", 2, vts);
			printf("Tracks: %*" PRIu16 ", ", 2, dvd_vts[vts].tracks);
			printf("Valid: %*" PRIu16 ", ", 2, dvd_vts[vts].valid_tracks);
			printf("Invalid: %*" PRIu16, 2, dvd_vts[vts].invalid_tracks);
			printf("\n");
		}

	}

	// Display more specific Video Title Set information
	if(opt_vts) {
		printf("        Video Title Set: %*" PRIu16 ", ", 2, arg_vts);
		printf("Tracks: %*" PRIu16 ", ", 2, dvd_vts[arg_vts].tracks);
		printf("Valid tracks: %*" PRIu16 ", ", 2, dvd_vts[arg_vts].valid_tracks);
		printf("Invalid tracks: %*" PRIu16, 2, dvd_vts[arg_vts].invalid_tracks);
		printf("\n");
	}


	/** dvd_info output **/

	struct dvd_track dvd_track;
	struct dvd_video dvd_video;
	struct dvd_audio dvd_audio;
	struct dvd_subtitle dvd_subtitle;
	struct dvd_chapter dvd_chapter;
	struct dvd_cell dvd_cell;

	for(track_number = d_first_track; track_number <= d_last_track; track_number++) {

		dvd_track = dvd_tracks[track_number];
		dvd_video = dvd_tracks[track_number].dvd_video;

		// Skip if limiting to tracks with audio only
		if(d_has_audio && dvd_track.active_audio_streams == 0)
			continue;

		// Skip if limiting tracks to one with VOBSUB subtitles only (cc not supported)
		if(d_has_subtitles && dvd_track.active_subs == 0)
			continue;

		// Skip if limiting to a minimum # of seconds which the length doesn't meet
		if(opt_min_seconds && dvd_track.msecs < (arg_min_seconds * 1000))
			continue;

		// Skip if limiting to a minimum # of minutes which the length doesn't meet
		if(opt_min_minutes && dvd_track.msecs < (arg_min_minutes * 1000 * 60))
			continue;

		// Skip if limiting to one title set
		if(opt_vts && dvd_track.vts != arg_vts)
			continue;

		// Skip if limiting to valid only
		if(d_is_valid && (dvd_vts[dvd_track.vts].valid == false || dvd_track.valid == false))
			continue;

		// Need an IFO if checking for audio or sub track languages
		if(d_has_alang || d_has_slang) {
			if(dvd_vts[dvd_track.vts].valid == false) {
				fprintf(stderr, "VTS %" PRIu16 " is invalid, skipping\n", dvd_track.vts);
			} else {
				vts_ifo = ifoOpen(dvdread_dvd, dvd_track.vts);
			}
		}

		// Skip if audio track language stream isn't found
		if(d_has_alang && !dvd_track_has_audio_lang_code(vts_ifo, d_alang)) {
			ifoClose(vts_ifo);
			vts_ifo = NULL;
			continue;
		}

		// Skip if subtitle language stream isn't found
		if(d_has_slang && !dvd_track_has_subtitle_lang_code(vts_ifo, d_slang)) {
			ifoClose(vts_ifo);
			vts_ifo = NULL;
			continue;
		}

		// Close out for next round
		if(vts_ifo) {
			ifoClose(vts_ifo);
			vts_ifo = NULL;
		}

		// Display track information
		printf("Track: %*" PRIu16 ", ", 2, dvd_track.track);
		printf("Length: %s, ", dvd_track.length);
		printf("Chapters: %*" PRIu8 ", ", 2, dvd_track.chapters);
		printf("Cells: %*" PRIu8 ", ", 2, dvd_track.cells);
		printf("Audio streams: %*" PRIu8 ", ", 2, dvd_track.audio_tracks);
		printf("Subpictures: %*" PRIu8 ", ", 2, dvd_track.subtitles);
		printf("Title set: %*" PRIu16", ", 2, dvd_track.vts);
		printf("Filesize: % 5.0lf MBs", dvd_track.filesize_mbs);
		printf("\n");

		// Display video information
		if(d_video) {
			printf("        Video format: %s, ", dvd_video.format);
			printf("Aspect ratio: %s, ", dvd_video.aspect_ratio);
			printf("Width: %" PRIu16 ", ", dvd_video.width);
			printf("Height: %" PRIu16 ", ", dvd_video.height);
			printf("FPS: %s, ", dvd_video.fps);
			printf("Display format: %s", display_formats[dvd_video.df]);
			printf("\n");
		}

		// Display audio tracks
		if(d_audio && dvd_track.audio_tracks) {

			d_stream_num = 1;

			for(audio_track_ix = 0; audio_track_ix < dvd_track.audio_tracks; audio_track_ix++) {

				dvd_audio = dvd_track.dvd_audio_tracks[audio_track_ix];

				printf("        Audio: %*" PRIu8 ", ", 2, d_stream_num);
				printf("Language: %s, ", (strlen(dvd_audio.lang_code) ? dvd_audio.lang_code : "--"));
				printf("Codec: %s, ", dvd_audio.codec);
				printf("Channels: %" PRIu8 ", ", dvd_audio.channels);
				printf("Stream id: %s, ", dvd_audio.stream_id);
				printf("Active: %s", (dvd_audio.active ? "yes" : "no"));
				printf("\n");
				d_stream_num++;

			}

		}

		// Display subtitles
		if(d_subtitles && dvd_track.subtitles) {

			d_stream_num = 1;

			for(subtitle_track_ix = 0; subtitle_track_ix < dvd_track.subtitles; subtitle_track_ix++) {

				dvd_subtitle = dvd_track.dvd_subtitles[subtitle_track_ix];

				printf("        Subtitle: %*" PRIu8 ", ", 2, d_stream_num);
				printf("Language: %s, ", (strlen(dvd_subtitle.lang_code) ? dvd_subtitle.lang_code : "--"));
				printf("Stream id: %s, ", dvd_subtitle.stream_id);
				printf("Active: %s", (dvd_subtitle.active ? "yes" : "no"));
				printf("\n");

				d_stream_num++;

			}

		}

		// Display chapters
		if(d_chapters && dvd_track.chapters) {

			for(chapter_ix = 0; chapter_ix < dvd_track.chapters; chapter_ix++) {

				dvd_chapter = dvd_track.dvd_chapters[chapter_ix];
				printf("        Chapter: %*" PRIu8 ", ", 2, dvd_chapter.chapter);
				printf("Length: %s, ", dvd_chapter.length);
				printf("First cell: %*" PRIu8 ", ", 2, dvd_chapter.first_cell);
				printf("Last cell: %*" PRIu8 ", ", 2, dvd_chapter.last_cell);
				printf("Filesize: % 5.0lf MBs\n", dvd_chapter.filesize_mbs);

			}

		}

		// Display track cells
		if(d_cells && dvd_track.cells) {

			for(cell_ix = 0; cell_ix < dvd_track.cells; cell_ix++) {

				dvd_cell = dvd_track.dvd_cells[cell_ix];

				printf("        Cell: %*" PRIu8 ", ", 2, dvd_cell.cell);
				printf("Length: %s, ", dvd_cell.length);
				printf("First sector: %7" PRIu64 ", ", dvd_cell.first_sector);
				printf("Last sector: %7" PRIu64", ", dvd_cell.last_sector);
				printf("Filesize: % 5.0lf MBs", dvd_cell.filesize_mbs);
				printf("\n");

			}

		}

	}

	// Cleanup

	cleanup:

	if(vmg_ifo)
		ifoClose(vmg_ifo);

	if(vts_ifo)
		ifoClose(vts_ifo);

	if(dvdread_dvd)
		DVDClose(dvdread_dvd);

	return 0;

}
